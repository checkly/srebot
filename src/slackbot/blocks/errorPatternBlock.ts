import { ErrorClusterWithCount } from "../../db/error-cluster";

export const LIST_ERROR_PATTERNS_ACTION_ID = "list_error_patterns";

export function createErrorPatternsBlock(
  errorPatterns: ErrorClusterWithCount[],
) {
  errorPatterns.forEach((errorPattern) => {
    console.log(errorPattern.error_message);
  });
  return {
    blocks: [
      {
        text: {
          emoji: true,
          text: "Error Pattern - Top 20",
          type: "plain_text",
        },
        type: "header",
      },
      {
        type: "divider",
      },
      ...(errorPatterns.length > 0
        ? [
            {
              text: {
                text: "*(Count) Summary*\n*Pattern Details*",
                type: "mrkdwn",
              },
              type: "section",
            },
            ...errorPatterns.slice(0, 20).flatMap((errorPattern) => [
              {
                text: {
                  text: `*(${errorPattern.count}) ${errorPattern.error_message.split("\n")[0]}*\n\`\`\`${errorPattern.error_message.replaceAll('"', "")}\`\`\``,
                  type: "mrkdwn",
                },
                type: "section",
              },
              {
                type: "divider",
              },
            ]),
          ]
        : [
            {
              text: {
                text: "No error Pattern found",
                type: "plain_text",
              },
              type: "section",
            },
          ]),
    ],
  };
}
