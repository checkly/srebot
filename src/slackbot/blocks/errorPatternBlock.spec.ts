import { createErrorPatternsBlock } from "./errorPatternBlock";

describe("errorPatternBlock", () => {
  it("renders empty error patterns", () => {
    const blocks = createErrorPatternsBlock([]);
    expect(blocks).toMatchSnapshot();
  });

  it("renders single error pattern", () => {
    const blocks = createErrorPatternsBlock([
      {
        id: "123",
        error_message: "Error Pattern #1\nDetails of error pattern #1",
        count: 10,
        first_seen_at: new Date(),
        last_seen_at: new Date(),
        embedding: [],
        embedding_model: "model",
        account_id: "account1",
      },
    ]);
    expect(blocks).toMatchSnapshot();
  });

  it("renders multiple error patterns", () => {
    const blocks = createErrorPatternsBlock([
      {
        id: "123",
        error_message: "Error Pattern #1\nDetails of error pattern #1",
        count: 10,
        first_seen_at: new Date(),
        last_seen_at: new Date(),
        embedding: [],
        embedding_model: "model",
        account_id: "account1",
      },
      {
        id: "124",
        error_message: "Error Pattern #2\nDetails of error pattern #2",
        count: 20,
        first_seen_at: new Date(),
        last_seen_at: new Date(),
        embedding: [],
        embedding_model: "model",
        account_id: "account2",
      },
    ]);
    expect(blocks).toMatchSnapshot();
  });
});
