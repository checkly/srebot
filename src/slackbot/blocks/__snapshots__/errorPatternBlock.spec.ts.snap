// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorPatternBlock renders empty error patterns 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Error Pattern - Top 20",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "text": {
        "text": "No error Pattern found",
        "type": "plain_text",
      },
      "type": "section",
    },
  ],
}
`;

exports[`errorPatternBlock renders multiple error patterns 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Error Pattern - Top 20",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "text": {
        "text": "*(Count) Summary*
*Pattern Details*",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*(10) Error Pattern #1*
\`\`\`Error Pattern #1
Details of error pattern #1\`\`\`",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
    {
      "text": {
        "text": "*(20) Error Pattern #2*
\`\`\`Error Pattern #2
Details of error pattern #2\`\`\`",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
  ],
}
`;

exports[`errorPatternBlock renders single error pattern 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Error Pattern - Top 20",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "type": "divider",
    },
    {
      "text": {
        "text": "*(Count) Summary*
*Pattern Details*",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*(10) Error Pattern #1*
\`\`\`Error Pattern #1
Details of error pattern #1\`\`\`",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
  ],
}
`;
